### PR TITLE
`Programming exercises`: Fix the tooltip placement on the programming exercise timeline

### PR DESCRIPTION
**Pull request by danielbode**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).

#### Client
- [x] I added multiple screenshots/screencasts of my UI changes.

### Motivation and Context

The tooltips for the Timeline when creating a new Programming Exercise were sometimes displayed to the bottom and sometimes to the top. Also the text was only shown partially.

### Description

I fixed the issue https://github.com/ls1intum/Artemis/issues/5125.

### Steps for Testing

Prerequisites:
- 1 Course

1. Go to course
2. Click "Generate new Programming Exercise"
3. Scroll to Timeline
4. Check if the tooltips are now displayed correctly
5. Verify, that no other tooltip is broken now

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

_unchanged_

### Screenshots

video

### CODE REVIEW

**Review by Marvin-Vitcu — APPROVED** [review]
Tested on TS1. Works as expected.

Congrats for your very first PR :D

One thing maybe: the tooltip for "Release Date" and "Due Date" are both missing a dot at the end. Is it possible to add this in this commit as well?

Another thing, but this might be too much for this commit: change the background color of the tooltip to the grey color we use for the input in points, bonus points and submission policy. Right now the tooltip does not standout very much. Or even better: just leave the background color as is, but add a white border to it.

**Review by pantageepapa — APPROVED** [review]
Tested on ts1 and it works as expected! I agree with the previous comment on the dot after the sentence.

**Comment by danielbode** [comment]
> Tested on TS1. Works as expected.
> 
> Congrats for your very first PR :D
> 
> One thing maybe: the tooltip for "Release Date" and "Due Date" are both missing a dot at the end. Is it possible to add this in this commit as well?
> 
> Another thing, but this might be too much for this commit: change the background color of the tooltip to the grey color we use for the input in points, bonus points and submission policy. Right now the tooltip does not standout very much. Or even better: just leave the background color as is, but add a thin white border to it.

Thank you!
I added the dots for the mentioned tooltip texts.

I think editing the background color of the tooltips is out of the scope of this Issue as it will also affect all the other tooltips in the application. But I will add a new Issue for this.

**Review by Marvin-Vitcu — APPROVED** [review]

**Review by TheZoker — APPROVED** [review]
Tested on TS3, works as expected 👍

File: src/main/webapp/app/shared/components/help-icon.component.ts [hunk 1]
```diff
@@ -3,7 +3,7 @@ import { faQuestionCircle } from '@fortawesome/free-solid-svg-icons';
 
 @Component({
     selector: 'jhi-help-icon',
-    template: ` <fa-icon [icon]="faQuestionCircle" class="text-secondary" [placement]="placement" ngbTooltip="{{ text | artemisTranslate }}"></fa-icon> `,
+    template: ` <fa-icon [icon]="faQuestionCircle" class="text-secondary" [placement]="placement" container="body" ngbTooltip="{{ text | artemisTranslate }}"></fa-icon> `,
```
Comment (manuelmanso) [inline src/main/webapp/app/shared/components/help-icon.component.ts:6]: Could adding `container="body"` here break other tooltips? I suggest you add a step to the Testing steps indicating that we should look around the application for other tooltips and check if they are still working correctly.

**Review by manuelmanso — CHANGES_REQUESTED** [review]

File: src/main/webapp/app/shared/components/help-icon.component.ts (hunk 1 above)
Comment (danielbode) [reply src/main/webapp/app/shared/components/help-icon.component.ts:6]: `container="body"` sets the level, where the tooltip gets displayed on, to "body" (the top most level). The default behaviour is, that it is displayed on the level of the element, where it is attached to, but that leads to the tooltip being partially hidden when the div isn't big enough and the overflow is set to hidden. I think it is the expected behaviour of a tooltip to always be in front of anything else, so it shouldn't break anything.
But I think it is a good idea to verify that by testing and I will add it to the testing steps. I just went through all tooltip occurrences, that I could find, and I didn't see any breaking changes.

**Review by luisa-we — APPROVED** [review]
Tested on TS1, works as expected.

Just a small note: In the German version, all tooltips except "Veröffentlichungsdatum der Beispiellösung" are also missing the dot at the end. 
Personally, I would also leave out the brackets at the second sentence of "Automatische Tests", because I think it looks better. But that's just my opinion :)

File: src/main/webapp/app/shared/components/help-icon.component.ts (hunk 1 above)
Comment (Dominik-Weinzierl) [reply src/main/webapp/app/shared/components/help-icon.component.ts:6]: Using `container body` or changing the `z index` is best practice here. So i think it looks good :+1:

**Comment by danielbode** [comment]
> Tested on TS1, works as expected.
> 
> Just a small note: In the German version, all tooltips except "Veröffentlichungsdatum der Beispiellösung" are also missing the dot at the end. Personally, I would also leave out the brackets at the second sentence of "Automatische Tests", because I think it looks better. But that's just my opinion :)

I added the missing dots in the german texts.

**Review by luisa-we — APPROVED** [review]
Retested on TS1, everything works as expected now!

**Review by stefanwaldhauser — APPROVED** [review]
Small but good change :) Good Job 👍

File: src/main/webapp/app/shared/components/help-icon.component.ts (hunk 1 above)
Comment (manuelmanso) [reply src/main/webapp/app/shared/components/help-icon.component.ts:6]: Great, just wanted to be sure 👍

**Review by manuelmanso — DISMISSED** [review]
LGTM considering the previous request

**Review by ge65cer — CHANGES_REQUESTED** [review]
The rightmost tooltip is not fully displayed and overflows the browser window. 

Screenshot 2022-08-17 at 12 23 47

**Review by ge65cer — CHANGES_REQUESTED** [review]
The rightmost tooltip still overflows the browser window.

Screenshot 2022-08-31 at 11 03 02

**Comment by danielbode** [comment]
> The rightmost tooltip still overflows the browser window.
> 
> Screenshot 2022-08-31 at 11 03 02

I fixed it.

**Review by ge65cer — APPROVED** [review]
Tested locally, placement and layout is fixed.

Screenshot 2022-09-26 at 10 39 34

**Review by stefanwaldhauser — APPROVED** [review]
LGTM! GJ

**Pull request merged**